src/Graph.ts

```diff
@@ -453,7 +446,22 @@ export default class Graph {
 
 			module.dependencies.forEach(depModule => {
 				if (!depModule.isExternal) {
-					visit(<Module>depModule, seen);
+					if (depModule.id in parents) {
+						if (!allSeen[depModule.id]) {
+							let path = [depModule.id];
+							let id = module.id;
+							while (id !== depModule.id) {
+								path.push(id);
+								id = parents[id];
+								if (!id) break;
+							}
+							cycles.push({ id: depModule.id, path });
```

Comment:
I think we can now move all the logic into the cycle warning function as mentioned before.

Say - `this.warnCycles(id, parents)`, with the logic in there?

Would just be good to keep conceptual boundaries in place.

Reply:
`parents` have to be frozen at the time the cycle is detected since there can be a lot of parents per module. I moved the path logic to `warnCycle`, it's not as nice as it was here imho, but probably worth it.

Reply:
The last patch doesn't correctly work. Sometimes the cycle doesn't end where it started.